### PR TITLE
Upgrade: fix replica agreements

### PR DESCRIPTION
**Pull request by flo-renaud**
### Upgrade: fix replica agreement

The upgrade checks the replication agreements to ensure that
some attributes are excluded from replication. The agreements
are stored in entries like
cn=serverToreplica,cn=replica,cn=_suffix_,cn=mapping tree,cn=config
but those entries are managed by the replication topology plugin
and should not be updated directly. The consequence is that the update
of the attributes fails and ipa-server-update prints an error message:

Error caught updating nsDS5ReplicatedAttributeList: Server is unwilling
to perform: Entry and attributes are managed by topology plugin.No direct
modifications allowed.
Error caught updating nsDS5ReplicatedAttributeListTotal: Server is
unwilling to perform: Entry and attributes are managed by topology
plugin.No direct modifications allowed.

The upgrade continues but the replication is not excluding
passwordgraceusertime.

Instead of editing the agreements, perform the modifications on
the topology segments.

Fixes: https://pagure.io/freeipa/issue/9385
Signed-off-by: Florence Blanc-Renaud <flo@redhat.com>

### Integration tests: add a test to ipa-server-upgrade

Add an integration test ensuring that the upgrade
properly updates the attributes to be excluded from
replication.

Related: https://pagure.io/freeipa/issue/9385
Signed-off-by: Florence Blanc-Renaud <flo@redhat.com>

### CODE REVIEW

File: ipatests/test_integration/test_simple_replication.py [hunk 1]
```diff
@@ -104,6 +105,24 @@ def test_ipa_custodia_check(self):
...
+        segment = find_segment(self.master, self.replicas[0], "domain")
+        self.master.run_command([
+            "ipa", "topologysegment-mod", "domain", segment,
+            "--replattrs",
+            "(objectclass=*) $ EXCLUDE memberof idnssoaserial entryusn"])
+        # Run the upgrade
+        result = self.master.run_command(["ipa-server-upgrade"])
+        # Ensure that the uprgade updated the attribute
+        errmsg = "Error caught updating nsDS5ReplicatedAttributeList"
+        assert errmsg not in result.stdout_text
```
Comment (rcritten) [inline ipatests/test_integration/test_simple_replication.py:124]: I think we can re-generate the expected attribute output as too to also ensure that the value itself was set in LDAP despite no error being returned. I realize this is a bit on the pedantic side but it's more of an affirmative test that the attribute was set to what we expect.

**Comment by rcritten** [comment]
The code change looks fine. I wonder if this bug has existed since we enabled the topology plugin (yikes).

File: ipatests/test_integration/test_simple_replication.py (hunk 1 above)
Comment (rcritten) [reply ipatests/test_integration/test_simple_replication.py:125]: Thanks, the test change looks good.

**Comment by flo-renaud** [comment]
master:

* 143c3eb1612f9bb7f015dcf2dcb496e8ef324a38 Upgrade: fix replica agreement
* ac78a84fbe90f361a4a58fb2748d539647ffea52 Integration tests: add a test to ipa-server-upgrade

**Pull request closed without merging**